### PR TITLE
Add support for groups defined by the #EXTGRP tag

### PR DESCRIPTION
**Pull request by arthur-liberman**

### CODE REVIEW

**Comment by phunkyfish** [comment]
Are you sure this doesn’t break anything? I have no way of testing it so if there are any follow on issues will you be able to pick them up?

**Comment by arthur-liberman** [comment]
I ran a few tests with my iptv playlist, and it seems to work ok. The groups are displayed correctly, channels are filtered properly according to the groups. Playback starts fine. If you have any suggestions what specifically to test I can do it.

**Comment by phunkyfish** [comment]
Not really. I don’t really remember the Leia version anymore so once you don’t mind picking up any issues it’s all good.

**Comment by phunkyfish** [comment]
Can you increment the middle number in addon xml and add a changelog entry?

**Comment by arthur-liberman** [comment]
I did some more testing, and it seems to work well with the following:
- Lists without any group markers
- Lists with `#EXTGRP` only
- Lists with `group-title=` only
- Lists with a mix of the above
- Lists with channels that have multiple groups + above (mixed)
- Lists with both `#EXTGRP` and `group-title=` on channels. In such case `group-title=` takes priority.

File: pvr.iptvsimple/addon.xml.in [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <addon
   id="pvr.iptvsimple"
-  version="3.9.8"
+  version="3.10.8"
```
Comment (phunkyfish) [inline pvr.iptvsimple/addon.xml.in]: Apologies. This should be 3.10.0. If we increment the minor number the micro number gets reset.

File: pvr.iptvsimple/changelog.txt [hunk 2]
```diff
@@ -1,3 +1,6 @@
+v3.10.8
```
Comment (phunkyfish) [inline pvr.iptvsimple/changelog.txt]: Same here

**Review by phunkyfish — CHANGES_REQUESTED** [review]
Just this small change and it looks good.

File: pvr.iptvsimple/addon.xml.in (hunk 1 above)
Comment (arthur-liberman) [reply pvr.iptvsimple/addon.xml.in]: Sure, I'll do that after work

File: src/PVRIptvData.cpp [hunk 3]
```diff
@@ -725,33 +727,16 @@ bool PVRIptvData::LoadPlayList(void)
...
-            }
-          }
-        }
+        iChannelGroupName = strGroupName;
+        ProcessGroupLine(strGroupName, bRadio, iCurrentGroupId);
       }
     }
+    else if (iChannelGroupName.empty() && StringUtils::Left(strLine, strlen(M3U_GROUP_MARKER)) == M3U_GROUP_MARKER)
+    {
+      iChannelGroupName = StringUtils::Right(strLine, (int)strLine.size() - strlen(M3U_GROUP_MARKER));
```
Comment (phunkyfish) [inline src/PVRIptvData.cpp]: C++ cast please. `static_cast<int>(strLine.size())`

File: src/PVRIptvData.cpp [hunk 4]
```diff
@@ -817,11 +802,11 @@ bool PVRIptvData::LoadPlayList(void)
         channel.strStreamURL = strLine;
       }
       
+      channel.strGroupName      = iChannelGroupName;
```
Comment (phunkyfish) [inline src/PVRIptvData.cpp]: Single space before and after `=`.

File: src/PVRIptvData.cpp [hunk 5]
```diff
@@ -861,6 +846,37 @@ bool PVRIptvData::LoadPlayList(void)
...
+  if (!groupsLine.empty())
+  {
+    std::stringstream streamGroups(groupsLine);
+    PVRIptvChannelGroup * pGroup;
+
+    while(std::getline(streamGroups, groupsLine, ';'))
+    {
+      groupsLine = XBMC->UnknownToUTF8(groupsLine.c_str());
+
+      if ((pGroup = FindGroup(groupsLine)) == NULL)
```
Comment (phunkyfish) [inline src/PVRIptvData.cpp]: `NULL` should be `nullptr`.

**Review by phunkyfish — CHANGES_REQUESTED** [review]

**Review by phunkyfish — APPROVED** [review]

**Comment by phunkyfish** [comment]
Thanks for the contribution!

**Comment by matthuisman** [comment]
@arthur-liberman 

getting quite a few reports of IPTV Simple getting stuck on "PVR Manager starting up" with 3.10.0 of IPTV Simple.
(it only made it into libreelec repo 10th feb which lots of my users use)

Users downgrading to 3.9.8 then have the issue fixed.
https://forums.whirlpool.net.au/thread/9x2zlv09?p=29#r561
https://mybroadband.co.za/forum/threads/release-dstv-now-kodi-add-on.977513/page-78#post-26966946

Looks like IPTV Simple loads up find when kodi starts.
But when it gets restarted, it hangs and ERROR: Process error processing job pvr-update-addons
I haven't personally been able to reproduce.
I'll try on a Pi4 LibreELEC today and hopefully can get it to happen.

Looks like the changes from 3.9.8 to 3.10.0 was just this commit:
https://github.com/kodi-pvr/pvr.iptvsimple/commit/fb1967a61f4e950ceea01fb9e651d654c79bda84

**just quickly looking at those changes - did   std::vector<int>::iterator it;   line 824 meant to be removed ?**
https://github.com/kodi-pvr/pvr.iptvsimple/commit/fb1967a61f4e950ceea01fb9e651d654c79bda84#diff-4c24ca87d9c14dd7ab6f9857835957e0b6026e40c1cb5e0b67a24a82f4b29a40L824
I'd imagine if not - that could be causing the crash...

**Comment by arthur-liberman** [comment]
@matthuisman Yes, I've seen a few reports of this, but I was not able to reproduce the issue.
By restart, do you mean disable/enable the addon?
We can restore this iterator, I'm just confused about it being there, as the for loop just after it uses its own iterator.
It's possible that some other change in the commit is responsible for this issue. Do we have any playlist samples that reproduce this problem?

**Comment by matthuisman** [comment]
i suspect it may be hardware related as I can't reproduce either on Windows.
I'm going to try on a RPI3 on LibreELEC shortly.
I don't think its related to the playlist as users of different playlists having the same.
And it looks like the error happens straight away on the 1st channel.

Yes, just disabling the add-on and re-enable is what IPTV Merge does to make it reload playlist.
So manually doing that should also pre-produce the issue
```
021-02-12 18:53:49.596 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: LoadPlayList PlayList Load Start
2021-02-12 18:53:49.597 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: Read line: '#EXTM3U'
2021-02-12 18:53:49.597 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: Read line: ''
2021-02-12 18:53:49.597 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: Read line: '#EXTINF:-1 channel-id="101" group-title="Entertainment;Foxtel Go" tvg-chno="101" tvg-id="F1S" tvg-logo="https://foxtel-go-sw.foxtelplayer.foxtel.com.au/go-mobile-440/imageHelper.php?id=3861118:206:CHANNEL:IMAGE:png&w=400#FOX One" tvg-name="FOX One",FOX One'
2021-02-12 18:53:49.597 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: Read line: 'plugin://plugin.video.foxtel.go/?_=play&id=3861118&media_type=1&_l=.pvr'
2021-02-12 18:53:49.597 T:3477308288   DEBUG: AddOnLog: PVR IPTV Simple Client: Found URL: 'plugin://plugin.video.foxtel.go/?_=play&id=3861118&media_type=1&_l=.pvr' (current channel name: 'FOX One', channel group: 'Entertainment;Foxtel Go')
2021-02-12 18:53:49.600 T:3477308288   ERROR: Process error processing job pvr-update-addons
```

**Comment by arthur-liberman** [comment]
Weird stuff, but looks like it just reproduced on one of my CE devices with some random playlist I had laying around. I'll try to bring back the iterator definition, and see if it fixes it or not.

**Comment by matthuisman** [comment]
great that you can reproduce.
 I'll be interested in what you find :)

**Comment by arthur-liberman** [comment]
Problem found, fixed. See #484 
Definitely a facepalm moment after I figured what I did wrong in this PR.
My apologies.

**Comment by matthuisman** [comment]
awesome. thank you for the quick solution :)

**Pull request merged**